teps/0059-skip-guarded-task-only.md

```diff
@@ -249,6 +287,263 @@ Users should be able to specify that a guarded `Task` only should be skipped whe
 - *ordering-dependent* `Tasks`, based on `runAfter`, should execute as expected
 - *resource-dependent* `Tasks`, based on resources such as `Results`, should be attempted but might be skipped if they can't resolve missing resources
 
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Today, we support specifying a list of `WhenExpressions` through the `when` field as such:
+
+```yaml
+when:
+  - input: 'foo'
+    operator: in
+    values: [ 'bar' ]
+```
+
+To provide the flexibility to skip a guarded `Task` when its `WhenExpressions` evaluate to `False` while unblocking the execution of its dependent `Tasks`, we propose changing the `when` field from a list to a dictionary and adding `scope` and `expressions` fields under the `when` field.
+- The `scope` field would be used to specify whether the `WhenExpressions` guard the `Task` only or the whole `Branch` (the `Task` and its dependencies). To unblock execution of subsequent `Tasks`, users would set `scope` to `Task`.
+- The `expressions` field would be used to specify the list of `WhenExpressions`, each of which has `input`, `operator` and `values` fields, as it is currently.
+
+
+```yaml
+when:
+  scope: Task
+  expressions:
+    - input: 'foo'
+      operator: in
+      values: [ 'bar' ]
+---
+when:
+  scope: Branch
+  expressions:
+    - input: 'foo'
+      operator: notin
+      values: [ 'bar' ]
+```
+
+To enable users to smoothly transition from the old syntax to the new syntax, we will initially support both sytanxes. Then we will deprecate the old syntax and support the new syntax only as we go towards v1 API.
```

Comment:
This means we will provide a default for `scope` — which will be `Branch` in order to not break backward compatibility, right ?
If so, any reason for, at some point (in v1) removing the default value ?

Reply:
@vdemeester no no, that means that we'll initially support both the current syntax and the proposed  syntax

current syntax:
```yaml
when:
  - input: 'foo'
    operator: in
    values: [ 'bar' ]
```

proposed syntax:
```yaml
when:
  scope: Task / Branch
  expressions:
    - input: 'foo'
      operator: in
      values: [ 'bar' ]
```

then only take the proposed syntax into v1 

the proposed syntax would not provide a default for `scope`, we'll require users to specify it because they have different expectation/intuition for the default behavior (https://github.com/tektoncd/community/pull/388) so being explicit about it ensures no one is surprised

I'd argue that this does not break backwards compatibility because the current syntax would continue working as is for a while until we stop supporting it later, what do you think?

/cc @bobcatfish

Reply:
oh I can't read properly it seems 😛 
Yes it makes sense to me now 😉